boundedint/int256.go

```diff
@@ -0,0 +1,138 @@
+
+// NewInt256 creates a Int256 struct with a minimum initial underlying value
+func NewInt256() *Int256 {
+	u := new(Int256)
+	u.value = *new(big.Int).Set(minInt256)
+	return u
+}
+
+// Value returns the underlying private value for a Int256 struct
+func (u *Int256) Value() big.Int {
+	return u.value
+}
+
+// Set assigns the underlying value of the given Int256 param to u, and returns the modified receiver struct
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Int256) Set(value big.Int) (*Int256, error) {
+	if value.Cmp(maxInt256) == 1 {
+		return nil, fmt.Errorf("cannot set Int256 to %v as it overflows max value of %v", value, maxInt256)
+	}
+	if value.Cmp(minInt256) == -1 {
+		return nil, fmt.Errorf("cannot set Int256 to %v as it underflows min value of %v", value, minInt256)
+	}
+	u.value = *new(big.Int).Set(&value)
+	return u, nil
+}
+
+// Copy sets the underlying value of u to a copy of the given Int256 param, and returns the modified receiver struct
+func (u *Int256) Copy(v *Int256) *Int256 {
+	u.value = *new(big.Int).Set(&v.value)
+	return u
+}
+
+// Cmp calls the underlying Cmp method for the big.Int stored in a Int256 struct as its value field
+func (u *Int256) Cmp(v *Int256) int {
+	return u.value.Cmp(&v.value)
+}
+
+// Equals returns true if the two Int256 structs have the same underlying values, false otherwise
+func (u *Int256) Equals(v *Int256) bool {
+	return u.Cmp(v) == 0
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the result falls outside of the signed 256-bit integer range
+func (u *Int256) Add(augend, addend *Int256) (*Int256, error) {
+	sum := new(big.Int).Add(&augend.value, &addend.value)
+	return u.Set(*sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the result falls outside of the signed 256-bit integer range
+func (u *Int256) Sub(minuend, subtrahend *Int256) (*Int256, error) {
+	difference := new(big.Int).Sub(&minuend.value, &subtrahend.value)
+	return u.Set(*difference)
+}
+
+// Mul sets u to multiplicand * multiplier and returns u as the product
+// returns an error when the result falls outside of the signed 256-bit integer range
+func (u *Int256) Mul(multiplicand, multiplier *Int256) (*Int256, error) {
+	product := new(big.Int).Mul(&multiplicand.value, &multiplier.value)
+	return u.Set(*product)
+}
+
+// String returns the string representation for Int256 structs
+func (u *Int256) String() string {
+	return u.value.String()
+}
+
+// MarshalJSON implements the json.Marshaler interface
+// it specifies how to marshal a Int256 struct so that it can be written to disk
+func (u Int256) MarshalJSON() ([]byte, error) {
+	return []byte(u.value.String()), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+// it specifies how to unmarshal a Int256 struct so that it can be reconstructed from disk
+func (u *Int256) UnmarshalJSON(b []byte) error {
```

Comment:
Are there tests for these encoding / decoding functions?

Reply:
yes. the `TestStore` function (`uint256_test.go:142`) covers this.